File: solutions/typescript/resistor-color-trio/2/resistor-color-trio.ts
export interface ResistanceWithUnit {
  name: string;
  divisor: number;
}

export interface ResistorValue {
  resistance: number;
  unit: ResistanceWithUnit;
}

export interface Unit {
  threshold: number;
  name: string;
  divisor: number;
}   

export const COLOR_SEQUENCE: string[] = [
  "black", "brown", "red", "orange", "yellow",
  "green", "blue", "violet", "grey", "white"
];  

export const UNITS: Unit[] = [
  { threshold: 9, name: "gigaohms", divisor: 1e9 },
  { threshold: 6, name: "megaohms", divisor: 1e6 },
  { threshold: 2, name: "kiloohms", divisor: 1e3 },
  { threshold: 0, name: "ohms", divisor: 1 }
];

export function decodedResistorValue(colors: string[]): string {
  const { resistance, unit } = resistanceWithUnit(colors);
  return `${resistance / unit.divisor} ${unit.name}`;
}

export function resistanceWithUnit(colors: string[]): ResistorValue {
  if (colors.length < 3) {
    throw new Error("At least three colors are required");
  }

  const firstValue = COLOR_SEQUENCE.indexOf(colors[0]);
  const secondValue = COLOR_SEQUENCE.indexOf(colors[1]);
  const multiplierExp = COLOR_SEQUENCE.indexOf(colors[2]);

  if (firstValue === -1 || secondValue === -1 || multiplierExp === -1) {
    throw new Error("Invalid color");
  }

  const baseValue = Number(`${firstValue}${secondValue}`);
  const resistance = baseValue * Math.pow(10, multiplierExp);

  const unit = UNITS.find(u => multiplierExp >= u.threshold)!;

  return { resistance, unit };
}

File: solutions/ruby/resistor-color-trio/3/resistor_color_trio.rb
class ResistorColorTrio
  CODING = {
    "black"  => 0, "brown"  => 1, "red"   => 2,
    "orange" => 3, "yellow" => 4, "green" => 5,
    "blue"   => 6, "violet" => 7, "grey"  => 8, "white"  => 9
  }.freeze

  UNITS = [
    { threshold: 9, name: "gigaohms", divisor: 10**9 },
    { threshold: 6, name: "megaohms", divisor: 10**6 },
    { threshold: 2, name: "kiloohms", divisor: 10**3 },
    { threshold: 0, name: "ohms",     divisor: 1 }
  ].freeze

  attr_reader :resistance, :unit_name, :divisor

  def initialize(colors)
    @resistance = calculate_resistance(colors)
    unit        = find_unit(colors[2])
    @unit_name  = unit[:name]
    @divisor    = unit[:divisor]
  end

  def label
    "Resistor value: #{resistance / divisor} #{unit_name}"
  end

  private

  def calculate_resistance(colors)
    first_two  = color_digit(colors[0]).to_s + color_digit(colors[1]).to_s
    multiplier = 10 ** color_digit(colors[2])
    
    first_two.to_i * multiplier
  end

  def find_unit(multiplier_color)
    exponent = color_digit(multiplier_color)
    
    UNITS.find { |u| exponent >= u[:threshold] }
  end

  def color_digit(color)
    CODING.fetch(color) do 
      raise ArgumentError, "Invalid color: #{color}"
    end
  end
end
